tests/components/songpal/test_config_flow.py

```diff
@@ -0,0 +1,249 @@
+"""Test the songpal config flow."""
+import copy
+
+from asynctest import MagicMock, patch
+from songpal import SongpalException
+from songpal.containers import InterfaceInfo
+
+from homeassistant.components import ssdp
+from homeassistant.components.songpal.const import CONF_ENDPOINT, DOMAIN
+from homeassistant.config_entries import SOURCE_IMPORT, SOURCE_SSDP, SOURCE_USER
+from homeassistant.const import CONF_HOST, CONF_NAME
+from homeassistant.data_entry_flow import (
+    RESULT_TYPE_ABORT,
+    RESULT_TYPE_CREATE_ENTRY,
+    RESULT_TYPE_FORM,
+)
+
+from tests.common import MockConfigEntry
+
+UDN = "uuid:1234"
+FRIENDLY_NAME = "friendly name"
+HOST = "0.0.0.0"
+ENDPOINT = f"http://{HOST}:10000/sony"
+MODEL = "model"
+
+SSDP_DATA = {
+    ssdp.ATTR_UPNP_UDN: UDN,
+    ssdp.ATTR_UPNP_FRIENDLY_NAME: FRIENDLY_NAME,
+    ssdp.ATTR_SSDP_LOCATION: f"http://{HOST}:52323/dmr.xml",
+    "X_ScalarWebAPI_DeviceInfo": {
+        "X_ScalarWebAPI_BaseURL": ENDPOINT,
+        "X_ScalarWebAPI_ServiceList": {
+            "X_ScalarWebAPI_ServiceType": ["guide", "system", "audio", "avContent"],
+        },
+    },
+}
+
+CONF_DATA = {
+    CONF_NAME: FRIENDLY_NAME,
+    CONF_ENDPOINT: ENDPOINT,
+}
+
+
+async def _async_return_value():
+    pass
+
+
+def _get_supported_methods(throw_exception):
+    def get_supported_methods():
+        if throw_exception:
+            raise SongpalException("Unable to do POST request: ")
+        return _async_return_value()
+
+    return get_supported_methods
+
+
+async def _get_interface_information():
+    return InterfaceInfo(
+        productName="product name",
+        modelName=MODEL,
+        productCategory="product category",
+        interfaceVersion="interface version",
+        serverName="server name",
+    )
+
+
+def _create_mocked_device(throw_exception=False):
+    mocked_device = MagicMock()
+    type(mocked_device).get_supported_methods = MagicMock(
+        side_effect=_get_supported_methods(throw_exception)
+    )
+    type(mocked_device).get_interface_information = MagicMock(
+        side_effect=_get_interface_information
+    )
+    return mocked_device
+
+
+def _patch_config_flow_device(mocked_device):
+    return patch(
+        "homeassistant.components.songpal.config_flow.Device",
+        return_value=mocked_device,
+    )
+
+
+def _flow_next(hass, flow_id):
+    return next(
+        flow
+        for flow in hass.config_entries.flow.async_progress()
+        if flow["flow_id"] == flow_id
+    )
+
+
+async def test_flow_ssdp(hass):
+    """Test working ssdp flow."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": SOURCE_SSDP}, data=SSDP_DATA,
+    )
+    assert result["type"] == "form"
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5462445)